Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      if (i == j) {
+        continue;
+      }
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+
+  // only merge digits from consecutive orbits
+  uint32_t orbit_p = previousBuffer.orbit;
+  uint32_t orbit_c = currentBuffer.orbit;
+  if (mPrint) {
+    std::cout << "orbit_c: " << orbit_c << "  orbit_p: " << orbit_p << std::endl;
+  }
+  if ((orbit_c >= orbit_p) && ((orbit_c - orbit_p) > 1)) {
+    return;
+  }
+
+  if (mPrint) {
+    std::cout << "Merging digits from " << currentBufId << " (orbit=" << currentBuffer.orbit << ") into "
+              << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < currentBuffer.digits.size(); i++) {
+    MergerDigit& d1 = currentBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+}
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].setOrbit(orbit, stop);
+}
+
+void Merger::setDigitHandler(std::function<void(const Digit&)> h)
+{
+  for (int feeId = 0; feeId <= MCH_MERGER_FEEID_MAX; feeId++) {
+    mergers[feeId].setDigitHandler(h);
+  }
+}
+
+void Merger::addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                      int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint8_t isStopRDH = 0;
+  uint32_t orbit;
+  uint32_t feeId;
+  uint32_t linkId;
+
+  const auto storeDigit = [&](const Digit& d) {
+    mOutputDigits.emplace_back(d);
+    if (mPrint) {
+      std::cout << "[storeDigit]: digit stored, mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    }
+  };
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
```

Comment:
could probably be replaced by a float chargeSum(const SampaCluster& sc) function in SampaCluster.h/.cxx ? btw what happens if this code is ran on data taken in chargesum mode ?

Reply:
Done. The code now also handles the chargeSum case properly.